proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,343 @@
+---
+simd: '0167'
+title: Loader-v4
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Review
+created: 2024-08-15
+feature: 2aQJYqER2aKyb3cZw22v4SL2xMX7vwXBRWfvS4pTrtED
+---
+
+## Summary
+
+A new upgradeable loader which only requires a single account per program.
+
+## Motivation
+
+Loader-v3, which is currently the only deployable loader, requires two accounts
+per program. And has three account types: Program, program-data and buffer.
+This was a workaround to circumvent the finality of the `is_executable` flag,
+which will be ignored by the program runtime from SIMD-0162 onwards.
+Consequentially, this setup of the program account being a proxy account,
+containing the address of the actual program data account, is no longer
+necessary and should be removed. Likewise the distinction of executable program
+accounts and non-executable buffer accounts is also no longer necessary and
+should be removed as well.
+
+In loader-v3 every instruction which modified the program data had to re-verify
+the ELF in the end. Instead we are now aiming for a more modular workflow which
+explicates these steps in a Retract-Modify-Redeploy sequence of instructions
+that allows multiple modifications to share one verification of the ELF in the
+end. This sequence can be a single transaction or split across multiple
+transactions.
+
+Another issue with loader-v3 is that the executable file stored in the
+programdata account is misaligned relative to the beginning of the account.
+This currently requires a copy in the ELF loader to re-align the program.
+
+Additionally, there currently is no complete specification of the loaders
+program management instructions. This proposal would thus fill that gap once
+loader-v4 goes into production.
+
+See impact for further motivation.
+
+## Alternatives Considered
```

Comment:
One thing I haven't been able to understand is why we added a new version of the loader instead of a new version of the program account state. Can you write up why we couldn't add a `UpgradeableLoaderState::ProgramV2` which doesn't have a programdata and is properly aligned? Seems like that would have been the simplest fix.

Reply:
Added

Reply:
Is it possible to add more details about the "special casing" that adding a new status makes it harder than a new loader? The existence of the `UpgradeableLoaderState` should have made this easier. Will loader v4 need "special casing" as well?